raiden-ts/src/transfers/epics/withdraw.ts

```diff
@@ -1,78 +1,370 @@
+        ),
+      );
+      // emit request once immediatelly, then wait until success, then retry every 30s
+      return retrySendUntil$(send, action$, notifier, httpTimeout);
+    }),
+  );
+
+/**
+ * Emits withdraw.failure if we get 'revealTimeout' blocks close to expiration
+ *
+ * Before this threshold we're still confident we can make a withdraw tx go through, and therefore
+ * give up if we get too close.
+ *
+ * @param action$ - Observable of withdrawRequest.request actions
+ * @param state$ - Observable of RaidenStates
+ * @param deps - Epics dependencies
+ * @param deps.config$ - Config observable
+ * @returns Observable of withdraw.failure actions
+ */
+export const withdrawExpiresSoonEpic = (
+  action$: Observable<RaidenAction>,
+  {}: Observable<RaidenState>,
+  { config$ }: RaidenEpicDeps,
+): Observable<withdraw.failure> =>
+  action$.pipe(
+    filter(withdrawMessage.request.is),
+    filter((action) => action.meta.direction === Direction.SENT),
+    withLatestFrom(config$),
+    mergeMap(([action, { revealTimeout }]) =>
+      action$.pipe(
+        filter(newBlock.is),
+        filter(({ payload }) => action.meta.expiration < payload.blockNumber + revealTimeout),
+        take(1),
+        mapTo(withdraw.failure(new RaidenError(ErrorCodes.WITHDRAW_EXPIRES_SOON), action.meta)),
+        // if withdrawMessage.success (WithdrawConfirmation) go through, don't fail
+        takeUntil(
+          action$.pipe(filter(isResponseOf<typeof withdrawMessage>(withdrawMessage, action.meta))),
+        ),
+      ),
+    ),
+  );
+
+/**
+ * Upon WithdrawConfirmation, send the on-chain withdraw request
+ *
+ * @param action$ - Observable of newBlock actions
+ * @param state$ - Observable of RaidenStates
+ * @param deps - Epics dependencies
+ * @param deps.address - Our address
+ * @param deps.signer - Signer instance
+ * @param deps.main - Main signer and address (if present)
+ * @param deps.log - Logger instance
+ * @param deps.getTokenNetworkContract - TokenNetwork contract getter
+ * @param deps.latest$ - Latest observable
+ * @returns Observable of withdrawExpired actions
+ */
+export const withdrawSendTxEpic = (
+  action$: Observable<RaidenAction>,
+  {}: Observable<RaidenState>,
+  { address, signer, main, log, getTokenNetworkContract, latest$ }: RaidenEpicDeps,
+): Observable<withdraw.success | withdraw.failure> =>
+  action$.pipe(
+    filter(withdrawMessage.success.is),
+    filter((action) => action.meta.direction === Direction.SENT),
+    concatMap((action) =>
+      latest$.pipe(
+        first(),
+        mergeMap(({ state, config }) => {
+          const { subkey: configSubkey, revealTimeout } = config;
+          // don't send on-chain tx if we're 'revealTimeout' blocks from expiration
+          // this is our confidence threshold when we can get a tx inside timeout
+          assert(
+            action.meta.expiration >= state.blockNumber + revealTimeout,
+            ErrorCodes.WITHDRAW_EXPIRES_SOON,
+          );
```

Comment:
Yes, it can. The former epic emits a `withdraw.failure` action to notify interested parts (so far, user and the retry send message epic) this withdraw already failed, and it has already failed because, even if a Confirmation message comes after that, caused by a retry or whatever, at this point the tx would be aborted anyway, so that epic just informs everyone it failed, and this assert makes sure it fails even if the message suddenly arrives. But good question.